src/gov-action-contracts/AIPs/AIPArbOS30/parent_contracts/AIPArbOS30UpgradeChallengeManagerAction.sol

```diff
@@ -0,0 +1,98 @@
+    ProxyAdmin public immutable govProxyAdmin;
+    address public immutable newChallengeManagerImpl;
+    IOneStepProofEntry public immutable osp;
+    bytes32 immutable condRoot;
+    IOneStepProofEntry public immutable condOsp;
+
+    constructor(
+        L1AddressRegistry _l1AddressRegistry,
+        bytes32 _newWasmModuleRoot,
+        ProxyAdmin _govProxyAdmin,
+        address _newChallengeManagerImpl,
+        IOneStepProofEntry _osp,
+        bytes32 _condRoot,
+        IOneStepProofEntry _condOsp
+    ) {
+        require(
+            Address.isContract(address(_l1AddressRegistry)),
+            "AIPArbOS30UpgradeChallengeManagerAction: l1AddressRegistry is not a contract"
+        );
+        require(
+            Address.isContract(address(_govProxyAdmin)),
+            "AIPArbOS30UpgradeChallengeManagerAction: _govProxyAdmin is not a contract"
+        );
+        require(
+            Address.isContract(_newChallengeManagerImpl),
+            "AIPArbOS30UpgradeChallengeManagerAction: _newChallengeManagerImpl is not a contract"
+        );
+
+        require(
+            Address.isContract(address(_osp)),
+            "AIPArbOS30UpgradeChallengeManagerAction: _osp is not a contract"
+        );
+
+        require(
+            Address.isContract(address(_condOsp)),
+            "AIPArbOS30UpgradeChallengeManagerAction: _condOsp is not a contract"
+        );
+        l1AddressRegistry = _l1AddressRegistry;
+        newWasmModuleRoot = _newWasmModuleRoot;
+        govProxyAdmin = _govProxyAdmin;
+        newChallengeManagerImpl = _newChallengeManagerImpl;
+        osp = _osp;
+        condRoot = _condRoot;
+        condOsp = _condOsp;
+    }
+
+    function peform() external {
+        // set new wasm module root
+        IRollupCore rollup = l1AddressRegistry.rollup();
+        IRollupAdmin(address(rollup)).setWasmModuleRoot(newWasmModuleRoot);
+
+        // verify:
+        require(
+            rollup.wasmModuleRoot() == newWasmModuleRoot,
+            "AIPArbOS30UpgradeChallengeManagerAction: wasm module root not set"
+        );
+
+        // set the new challenge manager impl
+        TransparentUpgradeableProxy challengeManager = TransparentUpgradeableProxy(
+            payable(address(l1AddressRegistry.rollup().challengeManager()))
+        );
+        govProxyAdmin.upgradeAndCall(
+            challengeManager,
+            newChallengeManagerImpl,
+            abi.encodeCall(IChallengeManagerUpgradeInit.postUpgradeInit, (osp, condRoot, condOsp))
+        );
+
+        // verify
+        require(
+            govProxyAdmin.getProxyImplementation(challengeManager) == newChallengeManagerImpl,
+            "AIPArbOS30UpgradeChallengeManagerAction: new challenge manager implementation set"
+        );
+        require(
+            IChallengeManagerUpgradeInit(address(challengeManager)).osp() == address(osp),
+            "AIPArbOS30UpgradeChallengeManagerAction: new OSP not set"
+        );
```

Comment:
we can have a similar check for the `cond` root, using the new `getOsp(bytes32)` getter